### PR TITLE
JIP-26 Voting Alternative Approach and Form Validation

### PR DESCRIPTION
**Pull request by jpfranci**
Adds on to dryu's PR, but instead just returns an `isUpvoted` and `isDownvoted` field whenever a post is retrieved from the server.
I do it this way because I found a bunch of jank race conditions from trying to cache everything on the client.

## Additional Changes
* Added toast notifications for uncaught/miscellaneous errors 
image
* Hooked up our login. create, and update user dialog to react hook form for validation purposes.
image
* Fix update user api to not specify the user being updated in the endpoint. Also made it update all username for those posts
* Take out todos that have been done

### CODE REVIEW

**Comment by srhsrhsrhsrh** [comment]
hmmm my changes broke but I'll figure it out tomorrow, I'm ok with it if everyone else is ok with it

File: packages/server/src/data/db/db-operations/post-ops.js [hunk 1]
```diff
@@ -117,68 +96,45 @@ const getVotedPostsByUserID = async (userId, upvote) => {
...
-  }
+  const filter = {
+    _id: postId,
+    upvoters: { $nin: [userId] },
+  };
 
-  const updatedPost = await Post.findByIdAndUpdate(postId, updateSpec, {
+  return Post.findOneAndUpdate(filter, updateSpec, {
     new: true,
+    projection: createProjectionObject(userId).$project,
```
Comment (dryu99) [inline packages/server/src/data/db/db-operations/post-ops.js:116]: ahhhhhhhhh 👌

File: packages/server/src/data/db/db-operations/post-ops.js [hunk 2]
```diff
@@ -85,27 +77,14 @@ const getPostsByUserID = async (userId, sortType) => {
...
+  aggregation.push(createProjectionObject(userId));
   aggregation.push({
     $sort:
       sortType === "popular"
         ? { score: -1, date: -1, _id: -1 }
         : { date: -1, _id: -1 },
   });
-  const posts = await Post.aggregate(aggregation);
-  return posts;
+  return Post.aggregate(aggregation);
```
Comment (dryu99) [inline packages/server/src/data/db/db-operations/post-ops.js:87]: I see the err of my ways now

**Review by dryu99 — APPROVED** [review]
Thanks for doing this jip feel free to merge whenever, I'll fix conflicts and merge `JIP-26-voting` into `main` right after but let me know if you want me to wait

**Comment by jpfranci** [comment]
@dryu99 This should have all of your changes already since I based out of your branch, I will just merge

**Pull request merged**